buildSrc/src/main/java/net/neoforged/neodev/NeoDevPlugin.java

```diff
@@ -647,7 +604,88 @@ static TaskProvider<CreateMinecraftArtifacts> configureMinecraftDecompilation(Pr
+            Project project,
+            Provider<Directory> neoDevBuildDir,
+            BinaryPatchDist dist,
+            NeoDevConfigurations neoDevConfigurations,
+            TaskProvider<CreateCleanArtifacts> createCleanArtifacts) {
+        var tasks = project.getTasks();
+
+        var distName = dist.name().toLowerCase(Locale.ROOT);
+        var capitalizedDist = StringUtils.capitalize(distName);
+        var binpatchesDir = neoDevBuildDir.map(dir -> dir.dir("artifacts/binpatches"));
+
+        var installerToolsConfig = neoDevConfigurations.getExecutableTool(Tools.INSTALLERTOOLS);
+        var baseJar = tasks.register("createBaseJar" + capitalizedDist, GenerateBaseJar.class, task -> {
+            task.setDescription("Generates the base jar for creating binary patches of the " + distName + " distribution");
+            task.setGroup(INTERNAL_GROUP);
+            if (dist == BinaryPatchDist.CLIENT || dist == BinaryPatchDist.JOINED) {
+                task.getMinecraft().from(createCleanArtifacts.flatMap(CreateCleanArtifacts::getRawClientJar));
+            }
+            if (dist == BinaryPatchDist.SERVER || dist == BinaryPatchDist.JOINED) {
+                task.getMinecraft().from(createCleanArtifacts.flatMap(CreateCleanArtifacts::getRawServerJar));
+            }
+            task.getMappings().set(createCleanArtifacts.flatMap(CreateCleanArtifacts::getClientMappings));
+            task.getOutput().set(binpatchesDir.map(dir -> dir.file(distName + "-base.jar")));
+            task.getNeoFormMappings().fileProvider(neoDevConfigurations.neoFormMappingsFiles.getIncoming().getArtifacts().getResolvedArtifacts()
+                    .map(artifacts -> artifacts.iterator().next().getFile()));
+            task.classpath(installerToolsConfig);
+        });
+
+        // Create the jar file in its target state. We will create binary patches to convert the base-jar to this jar.
+        var sourceSets = project.getExtensions().getByType(JavaPluginExtension.class).getSourceSets();
+        var mainSourceSet = sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME);
+        var clientSourceSet = sourceSets.getByName("client");
+
+        // We must extract the manifest from the base jar to disk, because Gradles Jar task cannot take the manifest
+        // content directly from another jar.
+        var extractManifest = tasks.register("extractBaseJarManifest" + capitalizedDist, ExtractManifest.class, task -> {
+            task.setDescription("Extract the manifest from the base " + distName + " jar to be able to set it as the manifest for the target jar.");
+            task.setGroup(INTERNAL_GROUP);
+            task.getJarFile().set(baseJar.flatMap(GenerateBaseJar::getOutput));
+            task.getManifest().set(binpatchesDir.map(dir -> dir.file(distName + "-manifest.mf")));
+        });
+        var modifiedJar = tasks.register("createModifiedJar" + capitalizedDist, Jar.class, task -> {
```

Comment:
I am not a huge fan of how this is done because the modified jar is actually the same for client and for joined. That should probably be made more explicit somehow.

Reply:
That is actually not correct, the joined jar has a different manifest.
Note: We should be very explicit here that client+server+joined are always treated as different "distributions" of the game, and we shouldn't try to be "smart" and reuse client jar for joined, etc.

Reply:
Changed to only diff classes, and as a result we use the same modified jar for client+joined now.